### PR TITLE
Use new charts

### PR DESCRIPTION
**Pull request by terrysahaidak**
Fixes RNBW-1891

### CODE REVIEW

**Comment by skylarbarrera** [comment]
took a look cause i was excited, performance seems good, not seeing any issues there which is 🔥, 
just a few kinks we should iron out.
- % change is 0 unless you are scrolling the graph: https://cloud.skylarbarrera.com/Screen-Recording-2021-11-30-15-54-45.mp4
- there is an initial animation when the chart loads, id imagine christian wont want 
- some of the spacing is off between the price &  % change on both platforms (edit: p sure this is preexisting so non blocking): https://cloud.skylarbarrera.com/Screen-Shot-2021-11-30-15-59-38.24.png

**Comment by christianbaroni** [comment]
> * there is a sometimes a flash with the timeseries where the options change, see 2nd chart: https://cloud.skylarbarrera.com/Screen-Recording-2021-11-30-16-03-34.mp4

this one's a preexisting issue but would love to fix

-----------

also noticed the high/low price labels are missing above and below the chart

aside from that and the issues skylar mentioned, this is great

**Comment by terrysahaidak** [comment]
Thank you guys for taking a look, gonna work on these issues

**Comment by christianbaroni** [comment]
hey terry this is looking great, just a few issues left

- when you first open a chart, the price label is often missing until you interact with the chart: https://www.dropbox.com/s/qdl18yin9158ljj/RPReplay_Final1638653737.mov?dl=0

- high/low labels should be equidistant from the chart path. there’s too much space between the low label + the path here, while the high label is correctly spaced: https://www.dropbox.com/s/mlnmjxo2t5dp3v4/IMG_0943.jpg?dl=0

- this one's an existing issue, but when you first open a chart, the transition from no chart → chart can be pretty rough. I wonder if we should use `ChartSpinner` before it shows up and then fade in the chart when it's loaded...possible the spinner is overkill and just a fade is cleaner. but ideally the sheet would open at full size without the two-step animation. and perhaps it would fix the jumping timespan labels (at 0:05 in the vid) if we waited till the chart is fully loaded to show it: https://www.dropbox.com/s/w31ivejuvt4gczt/RPReplay_Final1638606249.MOV?dl=0

**Comment by terrysahaidak** [comment]
Thanks for checking out @christianbaroni, seems like I need some iphone for testing too, price labels behave a bit different on Android and iphone 8 simulator (less space).

The missing price is only for etherium, seems to be related to the fact etherium chart data is preferched.

There is logic for the bottom sheet to "jump" to full height when the chart is loaded - should we expand it to the full height and show spinner instead?

File: src/components/value-chart/Chart.js
Comment (osdnk) [inline src/components/value-chart/Chart.js:200]: can you just move this outside of the function?

File: src/react-native-animated-charts/src/charts/linear/ChartLines.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartLines.tsx:54]: why 20?

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:277]: move this to some worklet, idk, i don't line copy-pasting

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:281]: shouldn't this be like:

```
{
	...timingFeedbackConfig,
	...timingFeedbackDefaultConfig
}
```
?

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:276]: can you for safety use platform-check?

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:81]: shouldn't this respect both? I think we typically want to use `smallestY, greatestY`, but when they're small we use the custom range? Didn't we have something similar before?

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:66]: can you move this outside of the function?

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:92]: and this? this doesn't need to be inside the component, right?

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:150]: how about useRef with a log to set the initial value only once? I don't this useMemo is really for this

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:196]: why not 
```
	return {
		...ctx,
		smallestX,
		smallestY,
		greatestX,
		greatestY 
	}
		
	}
```
?

File: src/react-native-animated-charts/src/helpers/requireOnWorklet.ts
Comment (osdnk) [inline src/react-native-animated-charts/src/helpers/requireOnWorklet.ts:38]:
```suggestion
    idRef.current = Date.now() * 10000 + Math.floor(Math.random() * 10000);
```

sounds crazy, but theoretically, I can think that module might collease

File: src/react-native-animated-charts/src/helpers/withReanimatedFallback.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/helpers/withReanimatedFallback.tsx:1]: ?

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:281]: The spread operator doesn't work in worklets. Previously, there was a function that would go recursively and merge them together. I don't see any point in it. If you want to use your own config - pass your own config. TS requires a compatible config. We can add this to README later.

File: src/react-native-animated-charts/src/charts/linear/ChartDot.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartDot.tsx:31]: why 10? the size?

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:281]: Why there's no point? I don't have a strong opinion, just curious

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:277]: I use the state shared value for compatibility reasons, I don't use it at all in the library. If I move this into a worklet - then it'll be copypasting of same worklet call which. Calling worklet is still a bit slower than not calling it.

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (osdnk) [reply src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:150]: According to docs, this might rerender extra time. 
Once, I've seen the pattern:

```
const initialPath = useState(() => { ... 
```

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:196]: I can use this too. My version is just a bit more "optimized" since we don't create yet another object where we copy all the stuff from ctx and third object. Instead, we just mutate ctx with new stuff. But for readability I can use your version i guess.

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:92]: Makes sense

File: src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPathProvider.tsx:81]: Here is the same behaviour as we had previously:
https://github.dev/rainbow-me/react-native-animated-charts/blob/1a623fbc5d6eca1a19c4ed37833c8d9677285ea8/src/charts/linear/ChartPath.js#L71-L72

File: src/react-native-animated-charts/src/charts/linear/ChartLines.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartLines.tsx:54]: I should've used the same comment as in the original code 😄 

```
// temporary fix for clipped chart
```

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:281]: under the hood Reanimated still merges our config with its default config. We don't provide anything special except for duration.

For spring most of the time you want to use all the values instead.

**Comment by osdnk** [comment]
According to what we talked with @christianbaroni, `ChartSpinner` will be part of the other PR. So extreme values are the only remaining things to fix.

**Comment by terrysahaidak** [comment]
Hey @christianbaroni,

Chart spinner will be addressed in a different PR since it's related to redux logic not to the charts itself - the fetchingCharts redux value is never true.

Also, about this one
>high/low labels should be equidistant from the chart path. there’s too much space between the low label + the path here, while the high label is correctly spaced: dropbox.com/s/mlnmjxo2t5dp3v4/IMG_0943.jpg?dl=0

Turned out this is the existing bug we had with the previous implementation. I think I know the solution, but it will take some more time to get it.

Since the whole point of this PR is to improve the performance of the initial render and overall I feel like it's been addressed by this PR and we can fix things above in later PRs.

File: src/react-native-animated-charts/src/helpers/requireOnWorklet.ts
Comment (osdnk) [inline src/react-native-animated-charts/src/helpers/requireOnWorklet.ts:7]: can you add a comment here why we do this?

**Review by osdnk — APPROVED** [review]
I read this code, run and I empathize with @terrysahaidak's point. This is good to go after @christianbaroni's approval and QA.

**Review by osdnk — APPROVED** [review]
I read this code, run and I empathize with @terrysahaidak's point. This is good to go after @christianbaroni's approval and QA.

**Review by christianbaroni — CHANGES_REQUESTED** [review]
found two issues testing on ios, otherwise working great

1. if you switch away from the 1D timeframe to any other timeframe, it should say "Past Week/Month/etc", but instead it shows a date: https://www.dropbox.com/s/r7vfp7s7w99l342/RPReplay_Final1638938348.mov?dl=0

2. the spacing in the chart was changed, here's an overlay on the old charts w some guides:
Group 1783

**Comment by christianbaroni** [comment]
once those are fixed, good to merge

**Comment by terrysahaidak** [comment]
Totally forgot about the overlay thing you mentioned on a call, gonna use it in the future, thanks @christianbaroni

**Comment by terrysahaidak** [comment]
The error on CI is related to types for d3-scale and d3-shape. These are only used in the charts library itself and present in its package.json. Not sure if I should include it in the package json of the Rainbow since they aren't actually used by it. But it won't pass the CI otherwise.

File: src/components/expanded-state/chart/ChartExpandedStateHeader.js
Comment (osdnk) [inline src/components/expanded-state/chart/ChartExpandedStateHeader.js:80]: shouldn't we use something `useImmediateEffect` here?

File: src/components/expanded-state/chart/chart-data-labels/ChartPercentChangeLabel.js
Comment (osdnk) [inline src/components/expanded-state/chart/chart-data-labels/ChartPercentChangeLabel.js:47]: so if `data` is falsy, then this is truth? Can this happen?


```suggestion
    data?.points?.length === 0 ?? true
```

File: src/components/value-chart/Chart.js
Comment (osdnk) [inline src/components/value-chart/Chart.js:233]: why did we increase this?

File: src/react-native-animated-charts/src/charts/linear/ChartDot.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartDot.tsx:27]: I think we're trying to use `React.FC`? I think this is the same as this, no?

```suggestion
const ChartDot = ({
  style,
  size = 10,
  springConfig,
  ...props
} : ChartDotProps) => {
```

File: src/react-native-animated-charts/src/charts/linear/ChartLines.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartLines.tsx:86]: maybe we can just get rid of `withReanimatedFallback`? We don't take advantage of this anyway.

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:22]: I think we're typically trying to used named imports:
```suggestion
import { getYForX } from 'react-native-redash';
```

File: src/react-native-animated-charts/src/charts/linear/ChartDot.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartDot.tsx:27]: No, React.FC also tells you that the function should return a valid React element as well as adds types for a children prop.

File: src/components/value-chart/Chart.js
Comment (terrysahaidak) [reply src/components/value-chart/Chart.js:233]: Oh this was for testing, good catch

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:351]:
```suggestion
          {...{ onGestureEvent }}
          onGestureEvent={onGestureEvent}
```

File: src/react-native-animated-charts/src/helpers/ChartContext.ts
Comment (osdnk) [inline src/react-native-animated-charts/src/helpers/ChartContext.ts:12]: we're not using `enum`s

```suggestion
export const CurveType = {
  basis: 'basis',
  bump: 'bump',
  linear: 'linear',
  monotone: 'monotone',
  natural: 'natural',
  step: 'step',
} as const;
```

File: src/components/value-chart/Chart.js
Comment (terrysahaidak) [reply src/components/value-chart/Chart.js:233]:
```suggestion
            <Dot color={colors.alpha(color, 0.03)} size={65}>
```

File: src/components/expanded-state/chart/chart-data-labels/ChartChangeDirectionArrow.js
Comment (terrysahaidak) [inline src/components/expanded-state/chart/chart-data-labels/ChartChangeDirectionArrow.js:33]:
```suggestion
  }, [colors]);
```

File: src/react-native-animated-charts/src/charts/linear/ChartLabels.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartLabels.tsx:19]:
```suggestion
        text: format?.(val.value) ?? val.value,
```
similarly, line below

File: src/react-native-animated-charts/src/charts/linear/ChartDot.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartDot.tsx:49]:
```suggestion
    const animation = withSpring(
        isActive.value ? 1 : 0,
        springConfig || springDefaultConfig
     )
      
    return {
      opacity: animation,
      transform: [
        { translateX },
        { translateY },
        {
          scale: animation,
        },
      ],
    };
```

File: src/react-native-animated-charts/src/charts/linear/ChartLines.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartLines.tsx:42]:
```suggestion
              (firstPoint?.y ?? 0) * height + 10 
```

typically, we use optional chaining, not ternary

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:246]:
```suggestion
          if (i === undefined || values.x === null) {
```

There's no chance that i is not defined is the scope, so we can compare this "normally" I think

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:250]: why hypot? Can you add some comment?

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:346]:
```suggestion
    const pathAnimatedStyles = useAnimatedStyle(() => ({
        opacity: pathOpacity.value * (1 - selectedOpacity) + selectedOpacity,
    }))
```

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:351]: why did this get resolved?

**Review by osdnk — APPROVED** [review]
Most of those comments are not important. Please, just add a comment in one place and include an explanation for using `hypot`.

**Comment by skylarbarrera** [comment]
only issue I am seeing rn is the price and time not showing up on initial render
https://cloud.skylarbarrera.com/Screen-Recording-2021-12-14-09-42-33.mp4

also somewhat curious, i know the issue was a mem leak but do we have any quantifiable data regarding perf improvements?

**Review by skylarbarrera — APPROVED** [review]
lgtm post CI passing

File: src/react-native-animated-charts/src/charts/linear/ChartLines.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartLines.tsx]: Done

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx]: Done

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:240]: I would use `typeof` here anyway since it's just a consistent thing with all of the checks for types.

File: src/components/expanded-state/chart/chart-data-labels/ChartChangeDirectionArrow.js
Comment (terrysahaidak) [reply src/components/expanded-state/chart/chart-data-labels/ChartChangeDirectionArrow.js:40]: It actually depends. But here is a good catch, since it won't work properly with change of theme.

File: src/components/expanded-state/chart/chart-data-labels/ChartPercentChangeLabel.js
Comment (terrysahaidak) [reply src/components/expanded-state/chart/chart-data-labels/ChartPercentChangeLabel.js:47]: I guess it's better to use just `!data?.points.length`

File: src/components/expanded-state/chart/ChartExpandedStateHeader.js
Comment (terrysahaidak) [reply src/components/expanded-state/chart/ChartExpandedStateHeader.js:80]: Probably, but I don't think it should be part of this PR

File: src/react-native-animated-charts/src/charts/linear/ChartLabels.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartLabels.tsx:19]: Hermes didn't work properly with these things. It's good it does. But tbh I don't like the cryptic look of `format?.()`

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:244]: I used this code as an example.
https://observablehq.com/@d3/multi-line-chart
I can't say you right now why or is there any better way of doing it, but it works :)

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:346]: I actually changed this already but seems like it's got missing

File: src/react-native-animated-charts/src/charts/linear/ChartLines.tsx
Comment (terrysahaidak) [reply src/react-native-animated-charts/src/charts/linear/ChartLines.tsx]: This code makes me overthink it and I don't like it. Ternary operator makes it way easier and obvious to understand what's going on here.

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:246]: I don't find those charts working correctly (the one you linked). I think this is really worth understanding because we want to have the responsibility for those charts in the future when we will want to use them

`Math.hypot` function returns the square root of the sum of squares of its arguments. In this case, we have one arg so $\sqrt{x^2} = \abs{x}$.
Why can't we just use `Math.abs`?
 
Also, `Math.floor(values.x)` is def smaller or eq to `currentPath.points[i].x`, so why won't we just remove the `abs` then?

 ```suggestion
          return currentPath.points[i].x - Math.floor(values.x)
```

If this will not work, I will be super worried.

Comment (osdnk) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:246]: As far as I understand, this will make more sense in R^3 space, but here I don't see what does this bring

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [inline src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:66]: As far as I understand, we use this function only in a single place for finding a point that is closest to some value. 

`compare` function returns a distance from `i` point to the one point over the x-axis.

image

So I expect the results will model this function with precision to qualification and a displacement over the x-axis.
We are willing to find a point that is the closes one to the "dip" (the root of this function when this touches the x-axis).

Why can't we use a divide-and-conquer bisection search for this? Complexity of this is `O(n)` where `n = length`.
Bisection can live us `O(log n)` that is nearly constant. 




```suggestion
function least(length: number, compare: typeof ascending = ascending) {
  'worklet';
  
  let bound1 = 0;
  let = length - 1;
  
  
  while(true) {
  	 let pivot = Math.round(bound1 + (bound2 - bound1) / 2)
  	 if (pivot === bound1) {
  	 	return bound1;
  	 } 
	  if (pivot === bound2) {
  	 	return bound2;
  	 }

  	 if (compare(pivot - 1) - compare(pivor) > 0) {
  	 	// decreasing, dip on the right side
  	 	bound1 = pivot;
  	 } else {
  	 	// non-increasing or dip, dip on the left side or in pivot
  	 	bound2 = pivot
  	 }
     }	
  }

```

File: src/react-native-animated-charts/src/charts/linear/ChartPath.tsx
Comment (osdnk) [reply src/react-native-animated-charts/src/charts/linear/ChartPath.tsx:246]: Okay, I am confused again `values.x` are values over x-axis? So `Math.floor(values.x)` isn't always `0`? If not, what are those `values.x`?

**Pull request merged**
